sentry_streams/sentry_streams/metrics/metrics.py

```diff
@@ -0,0 +1,236 @@
+        for name, value, tags in self.__timers.values():
+            self.datadog_client.timing(name.value, value, tags=tags)
+        for name, value, tags in self.__counters.values():
+            self.datadog_client.increment(name.value, value, tags=tags)
+        for name, value, tags in self.__gauges.values():
+            self.datadog_client.gauge(name.value, value, tags=tags)
+
+        self.__reset()
+
+    def __reset(self) -> None:
+        self.__timers.clear()
+        self.__counters.clear()
+        self.__gauges.clear()
+        self.__last_record_time = time.time()
+
+    def __throttled_record(self) -> None:
+        if time.time() - self.__last_record_time > METRICS_FREQUENCY_SEC:
+            self.flush()
+
+
+class ArroyoDatadogMetricsBackend:
+    """
+    Arroyo wrapper around Datadog metrics backend.
+    """
+
+    def __init__(self, datadog_client: DogStatsd) -> None:
+        self.__datadog_client = datadog_client
+
+    def __normalize_tags(self, tags: Tags) -> list[str]:
+        return [f"{key}:{value.replace('|', '_')}" for key, value in tags.items()]
+
+    def increment(
+        self,
+        name: ArroyoMetricName,
+        value: Union[int, float] = 1,
+        tags: Optional[Tags] = None,
+    ) -> None:
+        self.__datadog_client.increment(
+            name, value, tags=self.__normalize_tags(tags) if tags else None
+        )
+
+    def gauge(
+        self, name: ArroyoMetricName, value: Union[int, float], tags: Optional[Tags] = None
+    ) -> None:
+        self.__datadog_client.gauge(name, value, tags=self.__normalize_tags(tags) if tags else None)
+
+    def timing(
+        self, name: ArroyoMetricName, value: Union[int, float], tags: Optional[Tags] = None
+    ) -> None:
+        self.__datadog_client.timing(
+            name, value, tags=self.__normalize_tags(tags) if tags else None
+        )
+
+
+_metrics_backend: Optional[Metrics] = None
+_dummy_metrics_backend = DummyMetricsBackend()
+
+
+def configure_metrics(metrics: Metrics, force: bool = False) -> None:
+    """
+    Metrics can generally only be configured once, unless force is passed
+    on subsequent initializations.
+    """
+    global _metrics_backend
+
+    if not force:
+        assert _metrics_backend is None, "Metrics is already set"
+
+    # Perform a runtime check of metrics instance upon initialization of
+    # this class to avoid errors down the line when it is used.
+    assert isinstance(metrics, Metrics)
+    _metrics_backend = metrics
+    if isinstance(metrics, DatadogMetricsBackend):
+        arroyo_configure_metrics(ArroyoDatadogMetricsBackend(metrics.datadog_client))
+    else:
+        arroyo_configure_metrics(ArroyoDummyMetricsBackend())
+
+
+def get_metrics() -> Metrics:
+    if _metrics_backend is None:
+        return _dummy_metrics_backend
+    return _metrics_backend
+
+
+def get_size(obj: Any) -> int | None:
+    if isinstance(obj, (str, bytes)):
+        return len(obj)
+    return None
```

Comment:
What;s your goal here for the other message types? 

I'd recommend making `get_size` a method (or a property if you like it better) of the message class so each can implement it properly. That way we could also make messages rely on a header populated at the consumer level to track the size of the message. That would not impact anybody else.

If you plan to support other message types later, that's ok, in such case I would recommend adding a todo so we can track the missing work and still move the method to the message class now so we deal with the right interface.

Reply:
I tried the approach of using a `.size()` method but Reduce steps operate on arroyo messages not pipeline.Message. So I used this function instead. I'll add a TODO for the other work.